### PR TITLE
Display video in the Conversation area

### PR DESCRIPTION
**Pull request by gabebw**
Here's the flow:

* MainUI gets a token from `/api/token`, using Next's API Routes (https://nextjs.org/docs/api-routes/introduction)
  * The backend route is defined in `pages/api/token.ts`
* Once it has a token, it connects to a room
* Once it's connected to a room, it adds any existing participants and sets up hooks for detecting when new participants join (`participantConnected`) or existing participants leave (`participantDisconnected`)
* A `<Participant>` then renders each participant's video
  * The video track is rendered to a `<video>` tag
  * The audio track is rendered to `Audio()` (no HTML element) and hooked up to a `GainNode`

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# https://prettier.io/docs/en/options.html
+singleQuote: true
```
Comment (gabebw) [inline .prettierrc]: I love using Prettier. I added this to keep the style of what's already been built ❤️ since Prettier defaults to double-quotes by default.

File: components/MainUI.tsx [hunk 2]
```diff
@@ -1,13 +1,60 @@
...
+    setParticipants(previous =>
+      previous.filter(existing => existing.identity !== participant.identity)
+    );
+  };
+
+  const prepareRoom = room => {
+    setRoom(room);
+    setLocalParticipant(room.localParticipant);
+    window.addEventListener('beforeunload', () => room.disconnect());
+    [
```
Comment (gabebw) [inline components/MainUI.tsx]: This bit adds existing participants who were already in the room before you joined (`room.participants`).

File: components/MainUI.tsx [hunk 3]
```diff
@@ -1,13 +1,60 @@
...
+      ...Array.from(room.participants.values()),
+      room.localParticipant
+    ].forEach(participant => addParticipant(participant));
+    room.on('participantConnected', participant => {
```
Comment (gabebw) [inline components/MainUI.tsx]: Add a hook to detect participants who join after you. The hierarchy is:

* a room has participants
* participants have tracks (audio, video, and data)

File: components/MainUI.tsx [hunk 4]
```diff
@@ -1,13 +1,60 @@
...
+
+  useEffect(() => {
+    fetch('/api/token')
+      .then(res => res.json())
+      .then(data => {
+        Video.connectToRoom({ token: data.token, roomName: 'cool-people-only' })
+          .then(room => prepareRoom(room))
+          .catch(err => console.log(err));
+      });
+    return () => room && room.disconnect();
```
Comment (gabebw) [inline components/MainUI.tsx]: Returning this function means that this function will run when the component is unmounted. But wait, we're disconnecting above, too:

`window.addEventListener('beforeunload', () => room.disconnect());`

Yeah, true. I added the `beforeunload` handler because sometimes React doesn't have time to unload fully when reloading the page. We're covering all our bases here.

File: components/MainUI.tsx [hunk 5]
```diff
@@ -1,13 +1,60 @@
...
+  }, []);
+
   return (
     <LayoutContainer>
-      <Focus />
+      <Focus participants={participants} localParticipant={localParticipant}
```
Comment (gabebw) [inline components/MainUI.tsx]: It's a bit gross that we pass `participants` and `localParticipant` from `MainUI` ➡️ `Focus` ➡️ `Conversation`. We can refactor it to use context later to avoid this prop drilling.

File: components/Participant.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
+import React, { createRef, useEffect, useState } from 'react';
+
+const Participant = ({ gain, participant, isYou }) => {
+  const [videoRef] = useState(createRef<HTMLVideoElement>());
+
+  const playTrack = track => {
+    if (track.kind === 'audio') {
+      if (!isYou) {
+        // Don't play your own audio
```
Comment (gabebw) [inline components/Participant.tsx]: 😅

File: components/Participant.tsx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        var audio = new Audio();
+        audio.srcObject = postProcessAudioTrack(track);
+      }
+    } else if (track.kind === 'video') {
+      taggedLog('Creating new MediaStream');
+      taggedLog('track:', track);
+      const mediaStreamTrack = track.mediaStreamTrack;
+      const stream = new MediaStream();
+      stream.addTrack(mediaStreamTrack);
+      // @ts-ignore
```
Comment (gabebw) [inline components/Participant.tsx]: Ignored because Typescript doesn't know what `videoRef.current` refers to and errors out saying that you can't set `srcObject` on `Unknown`. ¯\_(ツ)_/¯

File: components/Participant.tsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      videoRef.current.srcObject = stream;
+    } else {
+      taggedLog('non-audio or video track:', track);
+    }
+  };
+
+  useEffect(() => {
+    participant.tracks.forEach(publication => {
```
Comment (gabebw) [inline components/Participant.tsx]: This is where the data track comes in, too. Right now we ignore it and only do something if `track.kind === 'audio'` or `track.kind === 'video'`.

File: components/Participant.tsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+    // hook it all up: source -> gain -> speakers
+    source.connect(gainNode).connect(audioCtx.destination);
+
+    return stream;
+  };
+
+  const taggedLog = (...args) => {
+    const identity = isYou ? 'you' : participant.identity;
+    console.log(...[`[${identity}]`, ...args]);
```
Comment (gabebw) [inline components/Participant.tsx]: This prefixes the arguments with the `identity` string, then passes it all to `console.log`. The net effect is this:

```js
// taggedLog("a", track, "b")
// is  equivalent to
console.log(`[${identity}]`, "a", track, "b")
```

Hooray for the array spread operator!

File: components/Video.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+const generateToken = () => {
+  const twilioAccountSid = process.env.TWILIO_SID;
+  const twilioApiKey = process.env.TWILIO_API_KEY;
+  const twilioApiSecret = process.env.TWILIO_API_SECRET;
+  const token = new jwt.AccessToken(
+    twilioAccountSid,
+    twilioApiKey,
+    twilioApiSecret
+  );
+  // @ts-ignore
```
Comment (gabebw) [inline components/Video.ts]: `token.identity` doesn't have a TS definition for some reason, so ignore it.

File: next.config.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+require('dotenv').config();
```
Comment (gabebw) [inline next.config.js]: Hey look, dotenv. Ask me for the `.env` file in Slack so you can run it locally.

File: components/Participant.tsx [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        audio.srcObject = postProcessAudioTrack(track);
+      }
+    } else if (track.kind === 'video') {
+      taggedLog('Creating new MediaStream');
+      taggedLog('track:', track);
+      const mediaStreamTrack = track.mediaStreamTrack;
+      const stream = new MediaStream();
+      stream.addTrack(mediaStreamTrack);
+      // @ts-ignore
+      videoRef.current.srcObject = stream;
```
Comment (eatspaint) [inline components/Participant.tsx]: You might try a [non-null assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator) here!
```suggestion
      videoRef.current!.srcObject = stream;
```

TLDR: It's a way to tell TypeScript "shut up, I know this isn't `undefined` || `null`"

File: components/Participant.tsx [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      taggedLog('non-audio or video track:', track);
+    }
+  };
+
+  useEffect(() => {
+    participant.tracks.forEach(publication => {
+      // sometimes it passes `null` (?!), so check if it's real
+      if (publication.track) {
+        playTrack(publication.track);
+      }
```
Comment (eatspaint) [inline components/Participant.tsx]:
```suggestion
    participant.tracks.forEach(({ track }) => {
        track && playTrack(track);
```

(I might have missed brackets in this suggestion, GH multi-line is weird)

File: components/Participant.tsx [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    participant.on('trackSubscribed', track => {
+      taggedLog(`trackSubscribed (${track.kind})`, track);
+      playTrack(track);
+    });
+  }, [participant, videoRef.current]);
+
+  const postProcessAudioTrack = audioTrack => {
+    // Manually add the stream so that we can post-process the audio
+    // @ts-ignore
+    const audioCtx = new (window.AudioContext || window.webkitAudioContext)();
```
Comment (eatspaint) [inline components/Participant.tsx]: I wonder if there's a polyfill (or if Babel has a plugin) for AudioContext/webkitAudioContext. I don't _love_ having to write/manage platform varied stuff like this :/

File: components/Video.ts (hunk 10 above)
Comment (eatspaint) [reply components/Video.ts]: You could try non-null assertion here as well!

File: components/Video.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+const generateToken = () => {
+  const twilioAccountSid = process.env.TWILIO_SID;
+  const twilioApiKey = process.env.TWILIO_API_KEY;
+  const twilioApiSecret = process.env.TWILIO_API_SECRET;
+  const token = new jwt.AccessToken(
+    twilioAccountSid,
+    twilioApiKey,
+    twilioApiSecret
+  );
```
Comment (eatspaint) [inline components/Video.ts]:
```suggestion
  const { TWILIO_SID, TWILIO_API_KEY, TWILIO_API_SECRET } = process.env;
  const token = new jwt.AccessToken(
    TWILIO_SID,
    TWILIO_API_KEY,
    TWILIO_API_SECRET,
  );
```

Also added a trailing comma in the args, that way if an additional one needs to be added for any reason, the diff won't have to touch the previous line!

File: components/Video.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+  // @ts-ignore
+  token.identity = `somebody-${getRandomIntInclusive(1, 100)}`;
```
Comment (eatspaint) [inline components/Video.ts]: You could use `Date.now()`(returns milliseconds since epoch) here to make it a little harder to have a collision!

**Comment by eatspaint** [comment]
Ok another stupid nit. The Zeit deploy tool (which you have to use to set secrets) [downcases the names of secrets](https://spectrum.chat/zeit/now/now-secrets-and-casing-issues~fdd085c7-5d6c-4e60-9382-22a3eb9c6491) 🙃. So we need to change our env secret names to be downcased too.

(the good news is I figured out how to set secrets, and Zeit pushes PR branches to review URLs which is REALLY COOL)

File: components/Video.ts (hunk 10 above)
Comment (gabebw) [reply components/Video.ts]: >You could try non-null assertion here as well!
 
What is that?

File: components/Conversation.tsx [hunk 17]
```diff
@@ -1,17 +1,40 @@
...
   grid-row: top / bottom;
   grid-column: left / right;
   ${flexCentered};
 `;
 
-const Conversation = () => {
+const Conversation = ({ localParticipant, participants }) => {
+  const isYou = participant => {
+    return localParticipant.identity === participant.identity;
+  };
```
[the remark this answers is not in the document]
Comment (gabebw) [reply components/Conversation.tsx]: Prettier is fine with it, but because the line is long, it turns into this, spread over 2 lines without braces:

```js
const isYou = participant =>
  localParticipant.identity === participant.identity;
```

What do you think?

File: components/MainUI.tsx [hunk 18]
```diff
@@ -1,13 +1,60 @@
...
 const { LayoutContainer } = Layout;
 
 const MainUI = () => {
+  const [participants, setParticipants] = useState([]);
+  const [room, setRoom] = useState(null);
+  const [localParticipant, setLocalParticipant] = useState(null);
+
+  const addParticipant = participant => {
+    console.log('Participant "%s" added', participant.identity);
+    setParticipants(previous => [...previous, participant]);
```
[the remark this answers is not in the document]
Comment (gabebw) [reply components/MainUI.tsx]: 🙇

File: components/MainUI.tsx [hunk 19]
```diff
@@ -1,13 +1,60 @@
...
+      previous.filter(existing => existing.identity !== participant.identity)
+    );
+  };
+
+  const prepareRoom = room => {
+    setRoom(room);
+    setLocalParticipant(room.localParticipant);
+    window.addEventListener('beforeunload', () => room.disconnect());
+    [
+      ...Array.from(room.participants.values()),
```
[the remark this answers is not in the document]
Comment (gabebw) [reply components/MainUI.tsx]: Yeah, it's not an array, it's a [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map).

File: components/Video.ts (hunk 10 above)
Comment (gabebw) [reply components/Video.ts]: Oh, I see your comment below. Thanks!

Comment (gabebw) [reply components/Video.ts]: Hmm, the error is:

>2339: Property 'identity' does not exist on type 'AccessToken'.

I'm not sure that a non-null assertion will fix it?

File: components/Video.ts (hunk 15 above)
Comment (gabebw) [reply components/Video.ts]: Usually I love the trailing comma, but Prettier removed the trailing comma when I added it 😬

Comment (gabebw) [reply components/Video.ts]: Great call on the destructuring of `process.env`.

File: components/Video.ts (hunk 16 above)
Comment (gabebw) [reply components/Video.ts]: Ah, I like having something human-distinguishable at a glance (45 vs 79, rather than 1584936034273 vs 1584936045323).

I think in the future we should just ask people for their name.

File: components/Participant.tsx (hunk 14 above)
Comment (gabebw) [reply components/Participant.tsx]: Yeah, I agree. I think we can just write our own "polyfill":

```js
const AudioContext = window.AudioContext || window.webkitAudioContext;
```

In the future, I imagine us having a centralized `Audio` class that we do all this work from, rather than spreading it all over random React components like I did here. So the grossness will at least be centralized later on.

**Comment by gabebw** [comment]
@eatspaint Thanks for the in-depth review! I implemented 90% of your feedback and asked questions about the other 10%.

File: components/MainUI.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (gabebw) [reply components/MainUI.tsx]: Or, to be more precise, `room.participants` is a Map and `room.participants.values()` is an Iterator.

File: .prettierrc (hunk 1 above)
Comment (WhittleGuy) [reply .prettierrc]: FWIW, my editor hates this file as soon as I open it, but putting the prettier config in the package.json as

`"prettier": {
      "singlequote": "true"
}`

was accepted without error. Good call though.

Comment (gabebw) [reply .prettierrc]: This file is in YAML format. If you set your editor to interpret all `.prettierrc` files as YAML, that should fix it.

File: components/Conversation.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (WhittleGuy) [reply components/Conversation.tsx]: IMO if it has to drop down (it looks to be one character too long), we should use curlies. That's easier to immediately parse if it ends up being multi-line regardless.

File: components/Participant.tsx [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+    const stream = new MediaStream();
+    stream.addTrack(audioTrack.mediaStreamTrack);
+    const source = audioCtx.createMediaStreamSource(stream);
+
+    // mute
+    const gainNode = audioCtx.createGain();
+    gainNode.gain.setValueAtTime(gain, audioCtx.currentTime);
+
+    // hook it all up: source -> gain -> speakers
+    source.connect(gainNode).connect(audioCtx.destination);
```
Comment (WhittleGuy) [inline components/Participant.tsx]: When/if this gets refactored to r-audio, all of this will be a nice lump of jsx, with a not-nice lump of dynamically set variables.

**Comment by gabebw** [comment]
okay, i'm merging this
